### PR TITLE
ci: Enable workflow_dispatch action smart-release

### LINKED ISSUE #2036: [FEATURE] One-button wash releases
### Affected project(s)

- [ ] documentation
- [ ] examples
- [ ] wasmCloud host
- [X] wasmCloud CLI (wash)
- [ ] wasmCloud dashboard UI (washboard)
- [ ] capability providers
- [ ] provider bindgen
- [ ] control interface client
- [ ] other / not sure

### Is your feature request related to a problem? Please describe.

`wash-cli` is a notoriously difficult crate to release with all kinds of inter-related wasmCloud dependencies.

### Describe the solution you'd like

A project like https://github.com/Byron/cargo-smart-release should be able to take the correct steps to update dependencies and publish new versions based on our conventional commits. I'm currently doing some experimentation here, but the idea here is that we could setup CI to be triggered on `workflow_dispatch` to run `cargo smart-release wash-cli`, which would compute the proper version of wash to release, and take a look at all of the downstream dependencies in this repository (namely, wash-lib, wasmcloud-control-interface, wasmcloud-core, provider-archive, wascap, the list goes on and on!) in order to release semver-proper releases of those libraries as well.

### Describe alternatives you've considered

- In the beginning, I think it would be great to use the output of cargo smart-release to manually publish tags, just in an ordered way, and use smart-release for confidence and changelogs.
- [cargo-release](https://github.com/crate-ci/cargo-release) is also a great tool that works with workspaces, but it doesn't automatically bump dependent crates
- [cargo-unleash](https://github.com/paritytech/cargo-unleash) also seems like a great tool, just hasn't had a commit in a little while
- [release-plz](https://crates.io/crates/release-plz) seems to work best with single crates, but combining this with actions to create the proper order of releases may work well too

I'd love to hear out alternatives here

### Additional context

https://github.com/brooksmtownsend/smart-releaser-demo is my repository where I'm testing out how this works with a top-level binary, and complex-ish dependencies on workspace crates.

### PR DESCRIPTION
**Pull request by brooksmtownsend**
## Feature or Problem
When releasing a crate or one of our main binaries, wash-cli or wasmcloud, currently it can be very difficult to properly predict:
1. What crates need to be released
2. What order crates need to be released
3. Whether to patch bump, minor bump, or leave as-is and release

Not only that, but we should be keeping up with a changelog as well. I chose to use cargo smart-release (with workspace-deps support https://github.com/Byron/cargo-smart-release/pull/17) in order to take care of the above actions automatically.

This PR adds a workflow_dispatch trigger to our `wasmcloud` action (if desired, I can migrate to a new action since I know that the workflow_dispatch wasn't release specific before) and lets you specify a crate to release. You can either directly release the crate, or leave the checkbox blank and the action will create a PR with the changes.

I attempted to capture the steps to use this action in https://github.com/wasmCloud/wasmCloud/wiki/Release-Runbook, which we should be able to keep updated so that all maintainers can reuse this for new releases.

## Related Issues
Fixes #2036 

## Release Information
N/A

### Manual Verification
You can see an example successful run of this in my fork:
1. Trigger action for `wasmcloud-core` without execute checked https://github.com/brooksmtownsend/wasmCloud/actions/runs/8973489750/job/24643786588
2. PR created: https://github.com/brooksmtownsend/wasmCloud/pull/3
3. PR merged, action triggered with execute checked
4. Action runs, creating tags and releases https://github.com/brooksmtownsend/wasmCloud/actions/runs/8973533429/job/24643936734, https://github.com/brooksmtownsend/wasmCloud/releases/tag/wasmcloud-core-v0.6.0

Additionally, by running directly on `main` with execute checked
1. Trigger action for `wasmcloud-tracing`, execute checked, all changes are reflected and pushed skipping 1,2,3 above https://github.com/brooksmtownsend/wasmCloud/actions/runs/8973902728/job/24645072545 (it _would_ pass but I want to avoid deleting old tags so that the above step shows fully what happened)

### CODE REVIEW

File: .github/workflows/wasmcloud.yml [hunk 1]
```diff
@@ -4,6 +4,20 @@ on:
...
+        type: choice
+        description: crate to smart-release
+        options:
+          - wasmcloud
+          - wasmcloud-core
+          - wasmcloud-tracing
+          - wash-cli
+          - wash-lib
+          - wascap
+      execute:
```
Comment (vados-cosmonic) [inline .github/workflows/wasmcloud.yml:18]: What about calling this `dry-run` or `pr-only` and `default`-ing it to `true`?

Or naming it `do-release`? or `publish-release`? `execute` just doesn't say much to me personally

File: .github/workflows/wasmcloud.yml [hunk 2]
```diff
@@ -655,3 +669,44 @@ jobs:
...
+      - uses: actions-rust-lang/setup-rust-toolchain@v1
+        with:
+          cache: false
+      - name: Install cmake for smart-release
+        run: sudo apt update && sudo apt install cmake -y
+
+      - name: install smart-release
+        env:
+          RUSTFLAGS: ''
+        run: cargo install cargo-smart-release --git https://github.com/brooksmtownsend/cargo-smart-release --branch feat/update-workspace-dependencies
```
Comment (vados-cosmonic) [inline .github/workflows/wasmcloud.yml:690]: Could you put in why we're using your branch version here? is that normal or was this just for testing?

File: .github/workflows/wasmcloud.yml [hunk 3]
```diff
@@ -655,3 +669,44 @@ jobs:
...
+      - name: dry run release
+        if: ${{ !inputs.execute }}
+        run: |
+          git config --global user.email "automation@wasmcloud.com"
+          git config --global user.name "wasmCloud automation"
+          cargo smart-release --update-crates-index --no-publish --execute --no-changelog-preview --no-push ${{ inputs.crate }}
```
Comment (vados-cosmonic) [inline .github/workflows/wasmcloud.yml:696]: NIT: maybe not worth doing, but we could decide the args to `smart-release` via a metadata step early on and then just *use* them from this step, enabling us to collapse dry run and regular release steps?

Could probably factor out the `git config` lines too

File: .github/workflows/wasmcloud.yml [hunk 4]
```diff
@@ -4,6 +4,20 @@ on:
...
+      crate:
+        type: choice
+        description: crate to smart-release
+        options:
+          - wasmcloud
+          - wasmcloud-core
+          - wasmcloud-tracing
+          - wash-cli
+          - wash-lib
+          - wascap
```
Comment (vados-cosmonic) [inline .github/workflows/wasmcloud.yml:17]: NIT: Could we get `wasmcloud-test-util` in here? That thing is mega old

**Review by vados-cosmonic — APPROVED** [review]
LGTM 🚀 nothing but nits

File: .github/workflows/wasmcloud.yml (hunk 2 above)
Comment (thomastaylor312) [reply .github/workflows/wasmcloud.yml:690]: Workspace dependency changes. It is in the PR description!

File: .github/workflows/wasmcloud.yml (hunk 3 above)
Comment (thomastaylor312) [reply .github/workflows/wasmcloud.yml:696]: I also would like this but not required

**Review by thomastaylor312 — APPROVED** [review]

File: .github/workflows/wasmcloud.yml (hunk 1 above)
Comment (brooksmtownsend) [reply .github/workflows/wasmcloud.yml:18]: Fair! Will noodle

File: .github/workflows/wasmcloud.yml (hunk 3 above)
Comment (brooksmtownsend) [reply .github/workflows/wasmcloud.yml:696]: I definitely think there are improvements here but can tackle as followups

File: .github/workflows/wasmcloud.yml (hunk 1 above)
Comment (brooksmtownsend) [reply .github/workflows/wasmcloud.yml]: Changed to `do-release`

Comment (vados-cosmonic) [reply .github/workflows/wasmcloud.yml]: Ahhh thank you for clarifying this, I'm definitely happy with that flow in the short to medium term

File: .github/workflows/wasmcloud.yml (hunk 2 above)
Comment (vados-cosmonic) [reply .github/workflows/wasmcloud.yml:693]: Oh yeah sorry a bit late here, but I was hoping to have that explanation *in the code*, appreciate the addition!

**Pull request merged**